### DISCUSSION TITLE
cluster Haproxy configuration

### DISCUSSION THREAD

**Discussion by nocpbxchap** [ant-media/Ant-Media-Server#5720] [General]
following these guide how we set origin and edge on configuration

https://antmedia.io/docs/guides/clustering-and-scaling/load-balancing/load-balancer-with-haproxy-ssl-termination/

also if using standalone database how we import to cluster database mongo?

regards

**Comment by tremzo** [comment]
If you are using standalone instance and want to convert to a cluster mode, simply use the following command:

`cd /usr/local/antmedia
sudo ./change_server_mode.sh cluster <MONGODB_SERVER_IP>`

Or if with credentials to MongoDB you an use:

`cd /usr/local/antmedia
sudo ./change_server_mode.sh cluster <MONGODB_SERVER_IP> <MONGODB_USERNAME> <MONGODB_PASSWORD>`

Please review the documentation here https://antmedia.io/docs/guides/clustering-and-scaling/manual-configuration/cluster-installation/#install-the-origin-and-edge-groups

**Reply by nocpbxchap** [reply]
yes correct i follow those steps, i am asking about haproxy

how we set origin server and edge server on haproxy configuration.

**Reply by tremzo** [reply]
In Haproxy document, there is no edge or origin configuration. They are going through the same backend instances. 

However, I will create an issue for that we'll update the documentation as soon as possible.

A single entry point:

```
frontend frontend_https
  bind *:443 ssl crt  /etc/haproxy/certs/{DOMAIN_NAME}.pem
  bind *:5443 ssl crt /etc/haproxy/certs/{DOMAIN_NAME}.pem
  reqadd X-Forwarded-Proto:\ https
  default_backend backend_http
```

load balancing multiple instances:

```
backend backend_http
  leastconn
  server ams1 {WRITE_YOUR_FIRST_AMS_SERVER_IP_ADDRESS}:5080  
  server ams2 {WRITE_YOUR_SECOND_AMS_SERVER_IP_ADDRESS}:5080 
```

**Comment by nocpbxchap** [comment]
i understand, then when i server reached capacity will go to second as
database is shared with al ams instances, right?

**Reply by tremzo** [reply]
In this configuration, the load is balanced between the ams instances defined in `backend backend_http` you'll need to add more instances if the amount of instances you have does not handle the load. Yes, in cluster mode, they all share the same MongoDB database.

**Comment by nocpbxchap** [comment]
with these configuration i have to put domain from principal server where i
publish or from load balancer server?

frontend frontend_https
  bind *:443 ssl crt  /etc/haproxy/certs/{DOMAIN_NAME}.pem
  bind *:5443 ssl crt /etc/haproxy/certs/{DOMAIN_NAME}.pem
  reqadd X-Forwarded-Proto:\ https
  default_backend backend_http

**Comment by muratugureminoglu** [comment]
Hi @nocpbxchap 

Here is a sample configuration, I have configured it for you :) I have not tested yet, FYI

```
frontend rtmp_lb
    bind *:1935 
    mode tcp
    backend_rtmp

backend backend_rtmp
    mode tcp
    server origin1 {WRITE_YOUR_FIRST_AMS_SERVER_IP_ADDRESS}:1935 check  # Ant Media Server instance 1
    server origin2 {WRITE_YOUR_SECOND_AMS_SERVER_IP_ADDRESS}:1935 check  # Ant Media Server instance 2
    # you can add more instances

frontend frontend_origin_https
    bind *:443 ssl crt  /etc/haproxy/certs/{DOMAIN_NAME}.pem
    reqadd X-Forwarded-Proto:\ https
    origin_backend_http

frontend frontend_edge_https
    bind *:5443 ssl crt /etc/haproxy/certs/{DOMAIN_NAME}.pem
    reqadd X-Forwarded-Proto:\ https
    edge_backend_http


backend origin_backend_http
    leastconn
    # below line forwards http requests to https, if you do not have SSL termination, remove it
    redirect scheme https if ! { ssl_fc }
    # below line provides session stickiness
    cookie JSESSIONID prefix nocache
    server origin1 {WRITE_YOUR_FIRST_AMS_SERVER_IP_ADDRESS}:5080 check cookie origin1  #if you do not use session stickiness, remove cookie ams1
    server origin2 {WRITE_YOUR_SECOND_AMS_SERVER_IP_ADDRESS}:5080 check cookie origin2  #if you do not use session stickiness, remove cookie ams2
    # you can add more instances


backend edge_backend_http
    leastconn
    # below line forwards http requests to https, if you do not have SSL termination, remove it
    redirect scheme https if ! { ssl_fc }
    # below line provides session stickiness
    cookie JSESSIONID prefix nocache
    server edge1 {WRITE_YOUR_FIRST_AMS_SERVER_IP_ADDRESS}:5080 check cookie edge1  #if you do not use session stickiness, remove cookie ams1
    server edge2 {WRITE_YOUR_SECOND_AMS_SERVER_IP_ADDRESS}:5080 check cookie edge2  #if you do not use session stickiness, remove cookie ams2
    # you can add more instances
```
